File: backend/routes/leaderboard.py
"""
routes/leaderboard.py
Supports Global, CP, and Hackathon leaderboards.
"""
from flask import Blueprint, request, jsonify, current_app
from flask_jwt_extended import jwt_required, get_jwt_identity
from models import db, User, Profile, HackathonResult
from utils.auth_middleware import roles_required

leaderboard_bp = Blueprint("leaderboard", __name__)


def get_lb_cache_key(current_user=None):
    """Generate a unique cache key based on query params and user college if scope is college."""
    args = request.args
    scope = args.get("scope", "global")
    
    # Crucial: Include the user's specific college domain if scope is college
    # to avoid different colleges sharing the same cache key.
    college_domain = ""
    if scope == "college" and current_user:
        try:
            college_domain = current_user.email.split("@")[1].strip().lower()
        except Exception:
            pass
            
    key = f"lb_{args.get('type','cp')}_{scope}_{college_domain}_{args.get('branch','')}_{args.get('year','')}"
    return key


@leaderboard_bp.route("/leaderboard", methods=["GET"])
@jwt_required()
def get_leaderboard():
    """
    GET /api/leaderboard?type=cp|hackathon|github|overall&scope=global|college
    """
    current_user_id = get_jwt_identity()
    current_user = User.objects(id=current_user_id).first()

    cache = current_app.cache
    cache_key = get_lb_cache_key(current_user)
    
    cached_data = cache.get(cache_key)
    if cached_data:
        return jsonify(cached_data), 200

    lb_type = request.args.get("type", "cp")
    scope = request.args.get("scope", "global") # Default to global to show all colleges
    branch = request.args.get("branch")
    year = request.args.get("year")

    # 1. Base User Query (Filtering)
    # Filter by role="student" to only show students on the leaderboard.
    user_query = User.objects(role="student")

    if scope == "college" and current_user:
        try:
            domain = current_user.email.split("@")[1].strip().lower()
            user_query = user_query.filter(email__endswith=f"@{domain}")
        except Exception:
            pass

    if branch:
        user_query = user_query.filter(branch=branch)
    if year:
        user_query = user_query.filter(year=int(year))

    # 2. Map LB Type to Sort Field
    sort_map = {
        "cp": "-cp_score",
        "hackathon": "-hackathon_score",
        "github": "-github_total_score",
        "overall": "-global_score",
        "global": "-global_score"
    }
    sort_field = sort_map.get(lb_type, "-cp_score")

    # 3. Query Profiles
    # Always query profiles bound to our student user_query.
    # This prevents orphaned profiles, limits entries to active student accounts,
    # and keeps the global and college scopes perfectly aligned.
    query = Profile.objects(user__in=user_query)
    query = query.order_by(sort_field)

    # 4. Pagination
    page = max(1, int(request.args.get("page", 1)))
    per_page = min(100, int(request.args.get("per_page", 50)))
    skip = (page - 1) * per_page

    if lb_type == "github":
        # Sort users using User.objects.order_by("-github_score")
        user_query = user_query.order_by("-github_score")
        total_count = user_query.count()
        users = user_query.skip(skip).limit(per_page)
        
        # Prefetch profiles for these users to optimize performance
        profiles_dict = {str(p.user.id): p for p in Profile.objects(user__in=users)}
        
        leaderboard = []
        for idx, user in enumerate(users):
            rank = skip + idx + 1
            p = profiles_dict.get(str(user.id))
            if not p:
                p = Profile(user=user)
                
            college_display = user.college
            if not college_display or college_display.strip().lower() == "unknown":
                try:
                    college_display = user.email.split("@")[1].strip().upper()
                except Exception:
                    college_display = "Unknown"

            entry = {
                "rank": rank,
                "user_id": str(user.id),
                "name": user.name,
                "branch": user.branch,
                "year": user.year,
                "college": college_display,
                "avatar_url": p.avatar_url,
                "github_url": p.github_url,
                "github_impl_score": user.github_implementation,
                "github_imp_score": user.github_impact,
                "github_work_score": user.github_working,
                "github_total_score": user.github_score,
                "github_review_reason": p.github_review_reason,
                "github_username": p.github_username or "",
                "github_rank": p.github_rank or "",
                "github_score": user.github_score,
            }
            leaderboard.append(entry)
    else:
        # 3. Query Profiles
        # Always query profiles bound to our student user_query.
        query = Profile.objects(user__in=user_query)
        query = query.order_by(sort_field)
        total_count = query.count()
        profiles = query.skip(skip).limit(per_page)
        leaderboard = []

        # Build Result
        for idx, p in enumerate(profiles):
            rank = skip + idx + 1
            user = p.user
            if not user:
                continue

            college_display = user.college
            if not college_display or college_display.strip().lower() == "unknown":
                try:
                    college_display = user.email.split("@")[1].strip().upper()
                except Exception:
                    college_display = "Unknown"

            entry = {
                "rank": rank,
                "user_id": str(user.id),
                "name": user.name,
                "branch": user.branch,
                "year": user.year,
                "college": college_display,
                "avatar_url": p.avatar_url,
            }

            if lb_type == "cp":
                entry.update({
                    "cf_handle": p.cf_handle,
                    "cf_rating": p.cf_rating,
                    "cf_rank": p.cf_rank,
                    "cf_problems_solved": p.cf_problems_solved,
                    "lc_username": p.lc_username,
                    "lc_rating": p.lc_rating,
                    "lc_problems_solved": p.lc_problems_solved,
                    "cp_score": round(p.cp_score, 1)
                })
            elif lb_type == "hackathon":
                h_count = HackathonResult.objects(user=user).count()
                entry.update({
                    "score": p.hackathon_score,
                    "hackathons_count": h_count
                })
            elif lb_type in ["overall", "global"]:
                entry.update({
                    "score": round(p.global_score, 1),
                    "global_score": round(p.global_score, 1),
                    "cp_score": round(p.cp_score, 1),
                    "hackathon_score": p.hackathon_score,
                    "github_score": round(user.github_score, 1) if user else 0.0
                })
            
            leaderboard.append(entry)

    response_data = {
        "leaderboard": leaderboard, 
        "total": total_count,
        "page": page,
        "per_page": per_page
    }
    cache.set(cache_key, response_data, timeout=300)

    return jsonify(response_data), 200


@leaderboard_bp.route("/hackathon/result", methods=["POST"])
@jwt_required()
@roles_required('admin', 'superadmin', 'reviewer')
def add_hackathon_result():
    """
    POST /api/hackathon/result
    (Manual entry by admins, students must use /api/hackathons/submit)
    """
    current_user_id = get_jwt_identity()
    user = User.objects(id=current_user_id).first_or_404()
    
    data = request.get_json()
    
    if user.role == 'student':
        target_user_id = current_user_id
    else:
        target_user_id = data.get("user_id", current_user_id)

    target_user = User.objects(id=target_user_id).first_or_404()
    h_name = data.get("hackathon_name")
    position = data.get("position", 0)
    points = data.get("points", 0)

    if not h_name:
        return jsonify({"error": "Missing hackathon_name"}), 400

    new_res = HackathonResult(
        user=target_user,
        hackathon_name=h_name,
        position=position,
        points=points
    )
    new_res.save()
    
    from utils.scoring import update_user_scores
    update_user_scores(target_user_id)

    return jsonify({"message": "Hackathon result added successfully", "result": new_res.to_dict()}), 201


@leaderboard_bp.route("/hackathon/result/<res_id>", methods=["DELETE"])
@jwt_required()
def delete_hackathon_result(res_id):
    """
    DELETE /api/hackathon/result/<id>
    """
    current_user_id = get_jwt_identity()
    res = HackathonResult.objects(id=res_id).first_or_404()
    
    if str(res.user.id) != current_user_id:
        return jsonify({"error": "Not authorized"}), 403
        
    res.delete()
    update_user_scores(current_user_id)
    
    return jsonify({"message": "Result deleted"}), 200

File: backend/src/routes/leaderboard.js
/**
 * routes/leaderboard.js
 * Supports Global, CP, Hackathon, GitHub, and Overall leaderboards with in-memory caching.
 */

import { Router } from 'express';
import NodeCache from 'node-cache';
import { verifyToken } from '../middleware/auth.js';
import { rolesRequired } from '../middleware/roles.js';
import { User, Profile, HackathonResult } from '../models/index.js';
import {
  getCurrentAcademicSession,
  calculateAdmissionYear,
  getCurrentYearOfStudy,
} from '../utils/academicYear.js';
import { updateUserScores, setCache } from '../utils/scoring.js';

const router = Router();
const cache = new NodeCache({ stdTTL: 300 }); // 5 minutes default TTL
setCache(cache);

function getLbCacheKey(req, currentUser) {
  const type = req.query.type || 'cp';
  const scope = req.query.scope || 'global';
  const branch = req.query.branch || '';
  const year = req.query.year || '';

  let collegeDomain = '';
  if (scope === 'college' && currentUser) {
    try {
      collegeDomain = currentUser.email.split('@')[1].trim().toLowerCase();
    } catch {
      // Ignore
    }
  }

  return `lb_${type}_${scope}_${collegeDomain}_${branch}_${year}`;
}

// ── GET /api/leaderboard ───────────────────────────────────────────
router.get('/leaderboard', verifyToken, async (req, res) => {
  try {
    const currentUser = await User.findById(req.userId);
    const cacheKey = getLbCacheKey(req, currentUser);

    const cachedData = cache.get(cacheKey);
    if (cachedData) return res.status(200).json(cachedData);

    const lbType = req.query.type || 'cp';
    const scope = req.query.scope || 'global';
    const branch = req.query.branch;
    const year = req.query.year;

    // Filter by role="student"
    const userQueryFilter = { role: 'student' };

    if (scope === 'college' && currentUser) {
      try {
        const domain = currentUser.email.split('@')[1].trim().toLowerCase();
        userQueryFilter.email = new RegExp(`@${domain}$`, 'i');
      } catch {
        // Ignore
      }
    }

    if (branch) userQueryFilter.branch = branch;

    // Reverse mapping for admission_year
    if (year) {
      const currentSession = getCurrentAcademicSession();
      if (year.toString().toLowerCase() === 'alumni' || parseInt(year, 10) > 4) {
        userQueryFilter.admission_year = { $lte: currentSession - 4 };
      } else {
        const parsedYear = parseInt(year, 10);
        if (!isNaN(parsedYear)) {
          userQueryFilter.admission_year = calculateAdmissionYear(parsedYear);
        }
      }
    }

    const page = Math.max(1, parseInt(req.query.page || '1', 10));
    const perPage = Math.min(100, parseInt(req.query.per_page || '50', 10));
    const skip = (page - 1) * perPage;

    let leaderboard = [];
    let totalCount = 0;

    if (lbType === 'github') {
      totalCount = await User.countDocuments(userQueryFilter);
      const users = await User.find(userQueryFilter)
        .sort({ github_score: -1 })
        .skip(skip)
        .limit(perPage);

      const userIds = users.map((u) => u._id);
      const profilesList = await Profile.find({ user: { $in: userIds } });
      const profilesMap = new Map();
      profilesList.forEach((p) => profilesMap.set(p.user.toString(), p));

      for (let idx = 0; idx < users.length; idx++) {
        const u = users[idx];
        const rank = skip + idx + 1;
        const p = profilesMap.get(u._id.toString()) || {};
        const yearInfo = getCurrentYearOfStudy(u.admission_year);

        let collegeDisplay = u.college;
        if (!collegeDisplay || collegeDisplay.trim().toLowerCase() === 'unknown') {
          try {
            collegeDisplay = u.email.split('@')[1].trim().toUpperCase();
          } catch {
            collegeDisplay = 'Unknown';
          }
        }

        leaderboard.push({
          rank,
          user_id: u._id.toString(),
          name: u.name,
          branch: u.branch,
          year: yearInfo.yearOfStudy,
          year_display: yearInfo.displayString,
          is_alumni: yearInfo.isAlumni,
          college: collegeDisplay,
          avatar_url: p.avatar_url || null,
          github_url: p.github_url || null,
          github_impl_score: u.github_implementation,
          github_imp_score: u.github_impact,
          github_work_score: u.github_working,
          github_total_score: u.github_score,
          github_review_reason: p.github_review_reason || '',
          github_username: p.github_username || '',
          github_rank: p.github_rank || '',
          github_score: u.github_score,
        });
      }
    } else {
      const studentUsers = await User.find(userQueryFilter).select('_id');
      const studentUserIds = studentUsers.map((u) => u._id);

      const sortMap = {
        cp: { cp_score: -1 },
        hackathon: { hackathon_score: -1 },
        overall: { global_score: -1 },
        global: { global_score: -1 },
      };
      const sortObj = sortMap[lbType] || { cp_score: -1 };

      const profileFilter = { user: { $in: studentUserIds } };
      totalCount = await Profile.countDocuments(profileFilter);

      const profiles = await Profile.find(profileFilter)
        .sort(sortObj)
        .skip(skip)
        .limit(perPage)
        .populate('user');

      for (let idx = 0; idx < profiles.length; idx++) {
        const p = profiles[idx];
        const u = p.user;
        if (!u) continue;

        const rank = skip + idx + 1;
        const yearInfo = getCurrentYearOfStudy(u.admission_year);
        let collegeDisplay = u.college;
        if (!collegeDisplay || collegeDisplay.trim().toLowerCase() === 'unknown') {
          try {
            collegeDisplay = u.email.split('@')[1].trim().toUpperCase();
          } catch {
            collegeDisplay = 'Unknown';
          }
        }

        const entry = {
          rank,
          user_id: u._id.toString(),
          name: u.name,
          branch: u.branch,
          year: yearInfo.yearOfStudy,
          year_display: yearInfo.displayString,
          is_alumni: yearInfo.isAlumni,
          college: collegeDisplay,
          avatar_url: p.avatar_url || null,
        };

        if (lbType === 'cp') {
          const dict = p.toDict();
          Object.assign(entry, {
            cf_handle: p.cf_handle || null,
            cf_rating: dict.final_codeforces_rating,
            actual_cf_rating: dict.actual_codeforces_rating,
            admin_cf_rating: dict.admin_codeforces_rating,

            lc_username: p.lc_username || null,
            lc_rating: dict.final_leetcode_rating,
            actual_lc_rating: dict.actual_leetcode_rating,
            admin_lc_rating: dict.admin_leetcode_rating,

            cc_username: dict.codechef_username || null,
            cc_rating: dict.final_codechef_rating,
            actual_cc_rating: dict.actual_codechef_rating,
            admin_cc_rating: dict.admin_codechef_rating,

            hr_username: dict.hackerrank_username || null,
            hr_rating: dict.final_hackerrank_rating,
            actual_hr_rating: dict.actual_hackerrank_rating,
            admin_hr_rating: dict.admin_hackerrank_rating,

            he_username: dict.hackerearth_username || null,
            he_rating: dict.final_hackerearth_rating,
            actual_he_rating: dict.actual_hackerearth_rating,
            admin_he_rating: dict.admin_hackerearth_rating,

            leaderboard_points: Math.round(p.cp_score * 100) / 100,
            cp_score: Math.round(p.cp_score * 100) / 100,
          });
        } else if (lbType === 'hackathon') {
          const hCount = await HackathonResult.countDocuments({ user: u._id });
          Object.assign(entry, {
            score: p.hackathon_score,
            hackathons_count: hCount,
          });
        } else if (['overall', 'global'].includes(lbType)) {
          Object.assign(entry, {
            score: Math.round(p.global_score * 10) / 10,
            global_score: Math.round(p.global_score * 10) / 10,
            cp_score: Math.round(p.cp_score * 10) / 10,
            hackathon_score: p.hackathon_score,
            github_score: Math.round((u.github_score || 0) * 10) / 10,
          });
        }

        leaderboard.push(entry);
      }
    }

    const responseData = {
      leaderboard,
      total: totalCount,
      page,
      per_page: perPage,
    };

    cache.set(cacheKey, responseData, 300);
    return res.status(200).json(responseData);
  } catch (err) {
    console.error('[GET /leaderboard] Error:', err.message);
    return res.status(500).json({ error: 'Internal server error' });
  }
});

// ── POST /api/hackathon/result ─────────────────────────────────────
router.post('/hackathon/result', verifyToken, rolesRequired('admin', 'superadmin', 'reviewer'), async (req, res) => {
  try {
    const currentAdmin = await User.findById(req.userId);
    const data = req.body || {};

    const targetUserId = currentAdmin.role === 'student' ? req.userId : (data.user_id || req.userId);
    const targetUser = await User.findById(targetUserId);
    if (!targetUser) return res.status(404).json({ error: 'Target user not found' });

    const hName = data.hackathon_name;
    const position = data.position || 0;
    const points = data.points || 0;

    if (!hName) return res.status(400).json({ error: 'Missing hackathon_name' });

    const newRes = await HackathonResult.create({
      user: targetUser._id,
      hackathon_name: hName,
      position,
      points,
    });

    await updateUserScores(targetUserId.toString());
    return res.status(201).json({
      message: 'Hackathon result added successfully',
      result: newRes.toDict(),
    });
  } catch (err) {
    console.error('[POST /hackathon/result] Error:', err.message);
    return res.status(500).json({ error: 'Internal server error' });
  }
});

// ── DELETE /api/hackathon/result/:res_id ───────────────────────────
router.delete('/hackathon/result/:res_id', verifyToken, async (req, res) => {
  try {
    const resId = req.params.res_id;
    const resultDoc = await HackathonResult.findById(resId);
    if (!resultDoc) return res.status(404).json({ error: 'Result not found' });

    if (resultDoc.user.toString() !== String(req.userId)) {
      return res.status(403).json({ error: 'Not authorized' });
    }

    await HackathonResult.deleteOne({ _id: resultDoc._id });
    await updateUserScores(req.userId.toString());

    return res.status(200).json({ message: 'Result deleted' });
  } catch (err) {
    console.error('[DELETE /hackathon/result/:res_id] Error:', err.message);
    return res.status(500).json({ error: 'Internal server error' });
  }
});

export default router;
